README.md
# Radix-Router
[![Build Status](https://travis-ci.org/SunDoge/radix-router.svg?branch=master)](https://travis-ci.org/SunDoge/radix-router)
[![Coverage Status](https://coveralls.io/repos/github/SunDoge/radix-router/badge.svg?branch=master)](https://coveralls.io/github/SunDoge/radix-router?branch=master)
[![crates.io](http://meritbadge.herokuapp.com/radix-router)](https://crates.io/crates/radix-router)
[![Released API docs](https://docs.rs/radix-router/badge.svg)](https://docs.rs/radix-router)

Radix-Router is a Rust port of [julienschmidt/httprouter](https://github.com/julienschmidt/httprouter).

## Usage
This is just a quick introduction.

Let's start with a `hello world` example:
```rust
extern crate futures;
extern crate hyper;
extern crate pretty_env_logger;
extern crate radix_router;

use futures::future;
use hyper::rt::{self, Future};
use hyper::service::service_fn;
use hyper::{Body, Request, Response, Server};
use radix_router::router::{BoxFut, Handler, Params, Router};
use std::sync::Arc;

fn index(_: Request<Body>, _: Params) -> BoxFut {
    let res = Response::builder().body("welcome!\n".into()).unwrap();
    Box::new(future::ok(res))
}

fn hello(_: Request<Body>, ps: Params) -> BoxFut {
    // let name = ps.by_name("name").unwrap();
    let name = &ps[0];
    let res = Response::builder()
        .body(format!("hello, {}!\n", name).into())
        .unwrap();
    Box::new(future::ok(res))
}

fn main() {
    pretty_env_logger::init();

    let addr = ([127, 0, 0, 1], 3000).into();
    let mut router: Router<Handler> = Router::new();
    router.get("/", Box::new(index));
    router.get("/hello/:name", Box::new(hello));
    let arc_router = Arc::new(router);
    // new_service is run for each connection, creating a 'service'
    // to handle requests for that specific connection.
    let new_service = move || {
        // This is the `Service` that will handle the connection.
        let router = arc_router.clone();
        service_fn(move |req| router.serve_http(req))
    };

    let server = Server::bind(&addr)
        .serve(new_service)
        .map_err(|e| eprintln!("server error: {}", e));

    println!("Listening on http://{}", addr);

    rt::run(server);
}
```

### Handler
The handler can be anything. You can store a `T` and get an `Option<&T>`. Notice that `&T` is immutable. We offer a default `radix_router::router::Handler` which can be a `fn` or `closure`. When using closure, you are able to capture outside parameters. For example:

```rust 
router.get("/", Box::new(get_echo));
router.post("/echo", Box::new(post_echo));
router.post("/echo/uppercase", Box::new(post_echo_uppercase));
router.post("/echo/reversed", Box::new(post_echo_reversed));
router.get("/some", Box::new(move |_, _| -> BoxFut {
    Box::new(future::ok(
        Response::builder().body(some_str.into()).unwrap(),
    ))
}));
```

### Named parameters
`:name` is a *named parameter*. The values are accessible via `Option<Params>`, which is a wrapped slice of `Param`s. You can get the value of a parameter either by its index in the slice. of by using the `by_name(name)` method.

Named parameters only match a single path segment:
```
Pattern: /user/:user

 /user/gordon              match
 /user/you                 match
 /user/gordon/profile      no match
 /user/                    no match
```

**Note:** Since this router has only explicit matches, you can not register static routes and parameters for the same path segment. For example you can not register the patterns `/user/new` and `/user/:user` for the same request method at the same time. The routing of different request methods is independent from each other.

### Catch-All parameters

The second type are *catch-all* parameters and have the form `*name`. Like the name suggests, they match everything. Therefore they must always be at the **end** of the pattern:

```
Pattern: /src/*filepath

 /src/                     match
 /src/somefile.go          match
 /src/subdir/somefile.go   match
```

### Static files
You can serve static files by using:
```rust
router.serve_files("/examples/*filepath", "examples");
```

## Examples
An echo server example is written. You can test it by running

```bash
$ cargo run --example echo
```

```bash
$ curl http://127.0.0.1:3000/echo
Try POSTing data to /echo

$ curl -d "param1=1&param2=2" -X POST http://127.0.0.1:3000/echo
param1=1&param2=2
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Coverage Status.

### Changed
- Optimize `Params`.
- Recommand using `Arc` to wrap `Router`.
- 2018 edition

### Removed
- `impl Service for Router`.
- `impl IntoFuture for Router`.

## [0.1.2] - 2018-07-28
### Changed
- Make `Router` attributes `pub`.
- `Handler` in README.

## [0.1.1] - 2018-07-18
### Added
- Docs badge.
- Readme link.
- `hello` example

## 0.1.0 - 2018-07-17
### Added
- `echo` example.
- Travis CI.
- Publish.

[Unreleased]: https://github.com/SunDoge/radix-router/compare/v0.1.2...HEAD
[0.1.2]: https://github.com/SunDoge/radix-router/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/SunDoge/radix-router/compare/v0.1.0...v0.1.1




src/path.rs
/// CleanPath is the URL version of path.Clean, it returns a canonical URL path
/// for p, eliminating . and .. elements.
///
/// The following rules are applied iteratively until no further processing can
/// be done:
///	1. Replace multiple slashes with a single slash.
///	2. Eliminate each . path name element (the current directory).
///	3. Eliminate each inner .. path name element (the parent directory)
///	   along with the non-.. element that precedes it.
///	4. Eliminate .. elements that begin a rooted path:
///	   that is, replace "/.." by "/" at the beginning of a path.
///
/// If the result of this process is an empty string, "/" is returned
pub fn clean_path(p: &str) -> String {
    // Turn empty string into "/"
    if p == "" {
        return "/".to_string();
    }

    let n = p.len();
    let mut buf: Vec<u8> = Vec::new();

    // Invariants:
	//      reading from path; r is index of next byte to process.
	//      writing to buf; w is index of next byte to write.

	// path must start with '/'

    let mut r = 1;
    let mut w = 1;

    if !p.starts_with("/") {
        r = 0;
        buf.resize(n + 1, 0);
        buf[0] = b'/';
    }

    let mut trailing = n > 1 && p.ends_with("/");
    let p = p.as_bytes();

    // A bit more clunky without a 'lazybuf' like the path package, but the loop
	// gets completely inlined (bufApp). So in contrast to the path package this
	// loop has no expensive function calls (except 1x make)

    while r < n {
        match p[r] {
            b'/' => r += 1,  // empty path element, trailing slash is added after the end
            b'.' => {
                if r + 1 == n {
                    trailing = true;
                    r += 1;
                } else if p[r + 1] == b'/' {
                    // . element
                    r += 2;
                } else if p[r + 1] == b'.' && (r + 2 == n || p[r + 2] == b'/') {
                    // .. element: remove to last /
                    r += 3;

                    if w > 1 {
                        // can backtrack
                        w -= 1;

                        if buf.is_empty() {
                            while w > 1 && p[w] != b'/' {
                                w -= 1;
                            }
                        } else {
                            while w > 1 && buf[w] != b'/' {
                                w -= 1;
                            }
                        }
                    }
                }
            }
            _ => {
                // real path element.
			    // add slash if needed
                if w > 1 {
                    buf_app(&mut buf, p, w, b'/');
                    w += 1;
                }

                // copy element
                while r < n && p[r] != b'/' {
                    buf_app(&mut buf, p, w, p[r]);
                    w += 1;
                    r += 1;
                }
            }
        }
    }

    // re-append trailing slash
    if trailing && w > 1 {
        buf_app(&mut buf, p, w, b'/');
        w += 1;
    }

    if buf.is_empty() {
        return String::from_utf8(p[..w].to_vec()).unwrap();
    }
    String::from_utf8(buf[..w].to_vec()).unwrap()
}

/// internal helper to lazily create a buffer if necessary
fn buf_app(buf: &mut Vec<u8>, s: &[u8], w: usize, c: u8) {
    if buf.is_empty() {
        if s[w] == c {
            return;
        }
        buf.resize(s.len(), 0);
        buf[..w].copy_from_slice(&s[..w]);
    }
    buf[w] = c;
}

#[cfg(test)]
mod tests {
    use super::*;

    // path, result
    fn clean_tests() -> Vec<(&'static str, &'static str)> {
        vec![
            // Already clean
            ("/", "/"),
            ("/abc", "/abc"),
            ("/a/b/c", "/a/b/c"),
            ("/abc/", "/abc/"),
            ("/a/b/c/", "/a/b/c/"),
            // missing root
            ("", "/"),
            ("a/", "/a/"),
            ("abc", "/abc"),
            ("abc/def", "/abc/def"),
            ("a/b/c", "/a/b/c"),
            // Remove doubled slash
            ("//", "/"),
            ("/abc//", "/abc/"),
            ("/abc/def//", "/abc/def/"),
            ("/a/b/c//", "/a/b/c/"),
            ("/abc//def//ghi", "/abc/def/ghi"),
            ("//abc", "/abc"),
            ("///abc", "/abc"),
            ("//abc//", "/abc/"),
            // Remove . elements
            (".", "/"),
            ("./", "/"),
            ("/abc/./def", "/abc/def"),
            ("/./abc/def", "/abc/def"),
            ("/abc/.", "/abc/"),
            // Remove .. elements
            ("..", "/"),
            ("../", "/"),
            ("../../", "/"),
            ("../..", "/"),
            ("../../abc", "/abc"),
            ("/abc/def/ghi/../jkl", "/abc/def/jkl"),
            ("/abc/def/../ghi/../jkl", "/abc/jkl"),
            ("/abc/def/..", "/abc"),
            ("/abc/def/../..", "/"),
            ("/abc/def/../../..", "/"),
            ("/abc/def/../../..", "/"),
            ("/abc/def/../../../ghi/jkl/../../../mno", "/mno"),
            // Combinations
            ("abc/./../def", "/def"),
            ("abc//./../def", "/def"),
            ("abc/../../././../def", "/def"),
        ]
    }

    #[test]
    fn test_path_clean() {
        let tests = clean_tests();
        for test in tests {
            let s = clean_path(test.0);
            assert_eq!(test.1, s);

            let s = clean_path(test.1);
            assert_eq!(test.1, s);
        }
    }

    // #[test]
    // fn test_path_clean_mallocs() {

    // }
}


examples/hello.rs
extern crate futures;
extern crate hyper;
extern crate pretty_env_logger;
extern crate radix_router;

use futures::future;
use hyper::rt::{self, Future};
use hyper::service::service_fn;
use hyper::{Body, Request, Response, Server};
use radix_router::router::{BoxFut, Handler, Params, Router};
use std::sync::Arc;

fn index(_: Request<Body>, _: Params) -> BoxFut {
    let res = Response::builder().body("welcome!\n".into()).unwrap();

    Box::new(future::ok(res))
}

fn hello(_: Request<Body>, ps: Params) -> BoxFut {
    // let name = ps.by_name("name").unwrap();
    let name = &ps[0];
    let res = Response::builder()
        .body(format!("hello, {}!\n", name).into())
        .unwrap();
    Box::new(future::ok(res))
}

fn main() {
    pretty_env_logger::init();

    let addr = ([127, 0, 0, 1], 3000).into();
    let mut router: Router<Handler> = Router::new();
    router.get("/", Box::new(index));
    router.get("/hello/:name", Box::new(hello));
    let arc_router = Arc::new(router);
    // new_service is run for each connection, creating a 'service'
    // to handle requests for that specific connection.
    let new_service = move || {
        // This is the `Service` that will handle the connection.
        let router = arc_router.clone();
        service_fn(move |req| router.serve_http(req))
    };

    let server = Server::bind(&addr)
        .serve(new_service)
        .map_err(|e| eprintln!("server error: {}", e));

    println!("Listening on http://{}", addr);

    rt::run(server);
}


examples/echo.rs
extern crate futures;
extern crate hyper;
extern crate pretty_env_logger;
extern crate radix_router;

use futures::future;
use hyper::rt::{self, Future, Stream};
use hyper::service::service_fn;
use hyper::{Body, Request, Response, Server};
use radix_router::router::{BoxFut, Handler, Params, Router};
use std::sync::Arc;

fn get_echo(_: Request<Body>, _: Params) -> BoxFut {
    // Box::new(future::ok(Response::new(Body::from("Try POSTing data to /echo"))))
    // *response.body_mut() = Body::from("Try POSTing data to /echo");
    let response = Response::builder()
        .body(Body::from("Try POSTing data to /echo"))
        .unwrap();
    Box::new(future::ok(response))
}

fn post_echo(req: Request<Body>, _: Params) -> BoxFut {
    // Box::new(future::ok(Response::new(req.into_body())))
    // *response.body_mut() = req.into_body();
    let response = Response::builder().body(req.into_body()).unwrap();
    Box::new(future::ok(response))
}

fn post_echo_uppercase(req: Request<Body>, _: Params) -> BoxFut {
    let mapping = req.into_body().map(|chunk| {
        chunk
            .iter()
            .map(|byte| byte.to_ascii_uppercase())
            .collect::<Vec<u8>>()
    });

    // *response.body_mut() = Body::wrap_stream(mapping);
    let response = Response::builder()
        .body(Body::wrap_stream(mapping))
        .unwrap();
    Box::new(future::ok(response))
}

fn post_echo_reversed(req: Request<Body>, _: Params) -> BoxFut {
    let reversed = req
        .into_body()
        .concat2()
        .map(move |chunk| {
            let body = chunk.iter().rev().cloned().collect::<Vec<u8>>();
            // *response.body_mut() = Body::from(body);
            // response
            Response::builder().body(Body::from(body)).unwrap()
        })
        .map_err(|e| e.into());
    Box::new(reversed)
}

fn main() {
    pretty_env_logger::init();

    let addr = ([127, 0, 0, 1], 3000).into();
    let some_str = "Some";

    // new_service is run for each connection, creating a 'service'
    // to handle requests for that specific connection.
    let mut router: Router<Handler> = Router::new();
    router.get("/", Box::new(get_echo));
    router.post("/echo", Box::new(post_echo));
    router.post("/echo/uppercase", Box::new(post_echo_uppercase));
    router.post("/echo/reversed", Box::new(post_echo_reversed));
    router.get(
        "/some",
        Box::new(move |_, _| -> BoxFut {
            Box::new(future::ok(
                Response::builder().body(some_str.into()).unwrap(),
            ))
        }),
    );
    router.serve_files("/examples/*filepath", "examples");
    let arc_router = Arc::new(router);

    let new_service = move || {
        // This is the `Service` that will handle the connection.
        // `service_fn_ok` is a helper to convert a function that
        // returns a Response into a `Service`.
        // service_fn_ok(|_| {
        //     Response::new(Body::from(PHRASE))
        // })

        // router.get("/some", |req, ps| {
        //     Box::new(future::ok(Response::new(Body::empty())))
        // });

        let router = arc_router.clone();
        service_fn(move |req| router.serve_http(req))
    };

    let server = Server::bind(&addr)
        .serve(new_service)
        .map_err(|e| eprintln!("server error: {}", e));

    println!("Listening on http://{}", addr);

    rt::run(server);
}
